File: app/src/main/java/com/testproject/kaera/ringtestapp/ui/TopListAdapter.java
package com.testproject.kaera.ringtestapp.ui;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.testproject.kaera.ringtestapp.R;
import com.testproject.kaera.ringtestapp.enteties.APIRedditItem;
import com.testproject.kaera.ringtestapp.ui.util.RecyclerViewAdapter;

import java.text.SimpleDateFormat;
import java.util.List;

public class TopListAdapter extends RecyclerViewAdapter<APIRedditItem, TopListAdapter.ViewHolder> {

    private SimpleDateFormat dateFormat = new SimpleDateFormat("dd MMM HH:mm");
    private static final String commentsPrefix = "%d comments";
    private static final String authorPrefix = "submitted by %s";

    private OnThumbnailClick thumbnailClickListener = item -> {
        // empty
    };

    public TopListAdapter() {
        super(null);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        return new ViewHolder(inflater.inflate(R.layout.row_top_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

//        @BindView(R.id.tv_title) TextView tvTitle;
//        @BindView(R.id.comment_count) TextView tvCommentCount;
//        @BindView(R.id.submitted_by) TextView tvSubmitted;
//        @BindView(R.id.time) TextView tvTime;
//        @BindView(R.id.thumbnail_image) ImageView imgThumb;

        private APIRedditItem item;

        public ViewHolder(View itemView) {
            super(itemView);
//            ButterKnife.bind(this, itemView);
        }

        void bind(APIRedditItem item) {
            this.item = item;
//            this.tvTitle.setText(item.getTitle());
//            this.tvCommentCount.setText(String.format(commentsPrefix, item.getCommentsCount()));
//            this.tvSubmitted.setText(String.format(authorPrefix, item.getAuthor()));
//            this.tvTime.setText(dateFormat.format(item.getCreatedDate()));
//            Picasso.get().load(item.getThumbnail()).fit().centerInside().into(this.imgThumb);
        }

//        @OnClick(R.id.thumbnail_image)
        public void onThumbnailClick() {
            thumbnailClickListener.onClick(item);
        }
    }

    public void setThumbnailClickListener(OnThumbnailClick thumbnailClickListener) {
        this.thumbnailClickListener = thumbnailClickListener;
    }

    public void setData(List<APIRedditItem> items) {
        this.items.clear();
        this.items.addAll(items);
        notifyDataSetChanged();
    }

    public interface OnThumbnailClick {

        void onClick(APIRedditItem item);
    }
}

File: app/src/main/java/com/testproject/kaera/ringtestapp/ui/TopListAdapter.kt
package com.testproject.kaera.ringtestapp.ui

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import com.testproject.kaera.ringtestapp.R
import com.testproject.kaera.ringtestapp.enteties.APIRedditItem
import com.testproject.kaera.ringtestapp.ui.TopListAdapter.ViewHolder
import com.testproject.kaera.ringtestapp.ui.util.RecyclerViewAdapter
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.row_top_item.*
import rx.functions.Action1
import java.text.SimpleDateFormat

class TopListAdapter : RecyclerViewAdapter<APIRedditItem, ViewHolder> {

    private val dateFormat = SimpleDateFormat("dd MMM HH:mm")
    private val commentsPrefix = "%d comments"
    private val authorPrefix = "submitted by %s"

    private lateinit var thumbnailClickAction: Action1<APIRedditItem>

    constructor() : super(null)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return ViewHolder(inflater.inflate(R.layout.row_top_item, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(override val containerView: View?) : RecyclerView.ViewHolder(containerView), LayoutContainer {

        private var item: APIRedditItem? = null

        internal fun bind(item: APIRedditItem?) {
            this.item = item
            tv_title.text = item!!.title
            comment_count.text = String.format(commentsPrefix, item.commentsCount)
            submitted_by.text = String.format(authorPrefix, item.author)
            time.text = dateFormat.format(item.createdDate)
            Picasso.get().load(item.thumbnail).fit().centerInside().into(thumbnail_image)
            thumbnail_image.setOnClickListener { thumbnailClickAction?.call(item) }
        }
    }

    fun setThumbnailClickListener(action: Action1<APIRedditItem>) {
        this.thumbnailClickAction = action
    }

    fun setData(items: List<APIRedditItem>) {
        this.items.clear()
        this.items.addAll(items)
        notifyDataSetChanged()
    }
}